cloudsync/smartsync.py

```diff
@@ -0,0 +1,374 @@
+            #     pc.get_latest()
+            if pc and pc[REMOTE].path != new_pc[REMOTE].path:
+                pcs.append(pc)
+                new_pc = pc
+            else:
+                done = True
+        pcs.reverse()
+        return pcs
+
+
+class SmartSyncState(SyncState):
+    """Enhances the syncstate to support smart syncing."""
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 storage: Optional['Storage'] = None,
+                 tag: Optional[str] = None,
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset: Set[SyncEntry] = set()
+        self.excludeset: Set[SyncEntry] = set()
+        self._callbacks: List[Callable] = list()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def register_auto_sync_callback(self, callback):
+        self._callbacks.append(callback)
+
+    def _smart_sync_ent(self, ent):
+        if ent[LOCAL].path and not self.providers[LOCAL].exists_path(ent[LOCAL].path):
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+            self.update_entry(ent, REMOTE, None, changed=True)
+        self.requestset.add(ent)
+        self.excludeset.discard(ent)
+
+    def smart_sync_path(self, remote_path) -> List[SyncEntry]:
+        # q: do we want to raise an exception if path isn't found?
+        # q: do we need to also sync path's parents all the way to the root?
+        #   a: probably, but it doesn't make sense to add them to the requestset, just sync them implicitly
+        #       or maybe it does make sense to add them to the requestset? that way they will get subscribed too
+        #       or maybe always sync all folders at all times, and only files are smartsynced
+        ents = self.lookup_path(REMOTE, remote_path)
+        for ent in ents:
+            self._smart_sync_ent(ent)
+        return ents
+
+    def smart_sync_oid(self, remote_oid) -> SyncEntry:
+        ent = self.lookup_oid(REMOTE, remote_oid)
+        self._smart_sync_ent(ent)
+        return ent
+
+    def _smart_unsync_ent(self, ent):
+        if ent[LOCAL].path:
+            ent_info = self.providers[LOCAL].info_path(ent[LOCAL].path)
+            if ent_info:
+                self.providers[LOCAL].delete(ent_info.oid)
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+        self.requestset.discard(ent)
+        self.excludeset.add(ent)
+
+    def _smart_unsync(self, ents, source_id) -> Optional[SyncEntry]:
+        if not ents:
+            raise FileNotFoundError(source_id)
+        for ent in ents:
+            if ent in self.requestset:
+                self._smart_unsync_ent(ent)
+                return ent
+        return None
+
+    def smart_unsync_ent(self, ent) -> Optional[SyncEntry]:
+        return self._smart_unsync([ent], ent)
```

Comment:
I feel like the `source_id` should just be the local path

Reply:
Or maybe the remote oid...

Reply:
source id is only for labelling purposes, so if the ent is the thing that triggered the smart_unsync than that is what we put for the label

Reply:
So I'm with that, I just feel that the whole sync entry is a little heavy and we could get away with just using the remote_oid or path as the source_id

Reply:
Heavy? This is only used as the parameter passed to the __init__ while creating an Exception to raise during an exception. Doesn't feel like a place to be stingy to me...

Reply:
Heavy not in terms of performance but readability of the logs. Is there a compelling reason that the entire SyncEntry needs to be in the logs for a FileNotFoundError?